frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/pushdialog/GitPushDialog.java

```diff
@@ -0,0 +1,358 @@
+    return pushController.getSelectedPushSpecs();
+  }
+
+  @Override
+  @UIEffect
+  public @Nullable JComponent getPreferredFocusedComponent() {
+    return listPanel.getPreferredFocusedComponent();
+  }
+
+  @Override
+  protected Action getOKAction() {
+    return pushAction;
+  }
+
+  @Override
+  @UIEffect
+  public void push(boolean forcePush) {
+    executeAfterRunningPrePushHandlers(
+        new Task.Backgroundable(project, /* title */ "Pushing...", /* canBeCancelled */ true) {
+          @Override
+          public void run(ProgressIndicator indicator) {
+            pushController.push(forcePush);
+          }
+        });
+  }
+
+  @Override
+  @UIEffect
+  public void executeAfterRunningPrePushHandlers(Task.Backgroundable activity) {
+    PrePushHandler.Result result = runPrePushHandlersInModalTask();
+    if (result == PrePushHandler.Result.OK) {
+      activity.queue();
+      close(OK_EXIT_CODE);
+    } else if (result == PrePushHandler.Result.ABORT_AND_CLOSE) {
+      doCancelAction();
+    } else if (result == PrePushHandler.Result.ABORT) {
+      // cancel push and leave the push dialog open
+    }
+  }
+
+  @UIEffect
+  public PrePushHandler.Result runPrePushHandlersInModalTask() {
+    FileDocumentManager.getInstance().saveAllDocuments();
+    AtomicReference<PrePushHandler.Result> result = new AtomicReference<>(PrePushHandler.Result.OK);
+    new Task.Modal(project, /* title */ "Checking Commits...", /* canBeCancelled */ true) {
+      @Override
+      public void run(ProgressIndicator indicator) {
+        result.set(pushController.executeHandlers(indicator));
+      }
+
+      @Override
+      @UIEffect
+      public void onThrowable(Throwable error) {
+        if (error instanceof PushController.HandlerException) {
+          PushController.HandlerException handlerException = (PushController.HandlerException) error;
+          Throwable cause = handlerException.getCause();
+
+          String failedHandler = handlerException.getFailedHandlerName();
+          java.util.List<String> skippedHandlers = handlerException.getSkippedHandlers();
+
+          String suggestionMessageProblem;
+          if (cause instanceof ProcessCanceledException) {
+            suggestionMessageProblem = "${failedHandler} has been cancelled.";
+          } else {
+            // PushController.HandlerException constructor guarantees that cause is not null
+            assert cause != null : "Exception cause is null";
+            super.onThrowable(cause);
+            suggestionMessageProblem = "${failedHandler} has failed. See log for more details.";
+          }
+
+          String suggestionMessageQuestion = skippedHandlers.isEmpty()
+              ? "Would you like to push anyway or cancel the push completely?"
+              : "Would you like to skip all remaining pre-push steps and push, or cancel the push completely?";
```

Comment:
Maybe for super-clarity/consistency with `okText`:
```suggestion
              : "Would you like to skip all remaining pre-push steps and push anyway, or cancel the push completely?";
```